phygnn/utilities/tf_utilities.py

```diff
@@ -19,3 +20,174 @@ def tf_log10(x):
     den = tf.math.log(tf.constant(10, dtype=num.dtype))
     return num / den
+
+
+def _mean_fill(x, mask):
+    """Fill NaN values in the input tensor with the mean of the non-NaN.
+    If all values are NaN, fill with 0. This is assumed to be either a 3D
+    or 4D tensor, with the trailing feature channel removed.
+
+    Parameters
+    ----------
+    x : tf.Tensor
+        Tensor with NaNs and shape (n_obs, spatial_1, spatial_2) or
+        (n_obs, spatial_1, spatial_2, n_temporal)
+    mask : tf.Tensor
+        Boolean mask of the non-NaN locations in the original x
+
+    Returns
+    -------
+    x : tf.Tensor
+        Input tensor with NaN values filled with the mean of the non-NaN
+        values or 0 if all values are NaN.
+    mask : tf.Tensor
+        Mask of the input tensor where True is not NaN and False is NaN.
+    """
+    not_nan_float = tf.cast(mask, x.dtype)
+
+    hi_res_zeroed = tf.where(mask, x, tf.zeros_like(x))
+
+    count = tf.reduce_sum(not_nan_float)
+    total = tf.reduce_sum(hi_res_zeroed)
+    mean = tf.math.divide_no_nan(total, count)
+
+    return tf.where(mask, x, mean)
+
+
+def mean_fill(x):
+    """Fill NaN values in the input tensor with the mean of the non-NaN.
+    If all values are NaN, fill with 0. This is assumed to be either a 4D
+    or 5D tensor, with the trailing feature channel included.
+
+    Parameters
+    ----------
+    x : tf.Tensor
+        Tensor with NaNs and shape (n_obs, spatial_1, spatial_2, n_features) or
+        (n_obs, spatial_1, spatial_2, n_temporal, n_features)
+
+    Returns
+    -------
+    x_filled : tf.Tensor
+        Input tensor with NaN values filled with the mean of the non-NaN
+        values or 0 if all values are NaN.
+    mask : tf.Tensor
+        Mask of the input tensor where 1 is not NaN and 0 is NaN.
+    """
+    mask = tf.math.logical_not(tf.math.is_nan(x))
+    hr_feat = [_mean_fill(x[..., i], mask[..., i]) for i in range(x.shape[-1])]
+    hr_feat = tf.stack(hr_feat, axis=-1)
+    return hr_feat, tf.cast(mask, x.dtype)
+
+
+def _idw_fill(x, coords, mask):
+    """IDW fill for flattened tensors without depth dimension. Assumes input
+    shape is (N), where N is the number of points (H * W).
+
+    Parameters
+    ----------
+    x : tf.Tensor
+        Tensor with NaNs and shape (N)
+    coords: tf.Tensor
+        Coordinates of the points in the image, shape (N, 2) where N is the
+        number of points (H * W)
+    mask : tf.Tensor
+        Mask of the input tensor where True is not NaN and False is NaN.
+
+    Returns
+    -------
+    x_filled: tf.Tensor
+        Flattened tensor with NaNs filled
+    """
+    nan_mask = tf.math.logical_not(mask)
+    valid_coords = tf.boolean_mask(coords, mask)
+    nan_coords = tf.boolean_mask(coords, nan_mask)
+    valid_vals = tf.boolean_mask(x, mask)
+
+    diffs = tf.expand_dims(nan_coords, 1) - tf.expand_dims(valid_coords, 0)
+    dists = tf.norm(diffs, axis=-1)
+
+    weights = 1 / dists
```

Comment:
Any reason dist could be zero? NSRDB IDW protects against this.

Reply:
No since the nan_coords and valid_coords never overlap